services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionAsyncClient.java

```diff
@@ -15,80 +15,131 @@
+                                                                          Function<T, CompletableFuture<ReturnT>> operation,
+                                                                          String bucketName,
+                                                                          CompletableFuture<ReturnT> returnFuture) {
+
+        // // TODO: will fix the casts with separate PR
+        ((S3AsyncClient) delegate()).headBucket(b -> b.bucket(bucketName)).whenComplete((response,
+                                                                                         throwable) -> {
+            if (throwable != null) {
+                if (isS3RedirectException(throwable)) {
+                    bucketToRegionCache.remove(bucketName);
+                    Optional<String> bucketRegion = getBucketRegionFromException((S3Exception) throwable.getCause());
+
+                    if (bucketRegion.isPresent()) {
+                        bucketToRegionCache.put(bucketName, Region.of(bucketRegion.get()));
+                        sendRequestWithRightRegion(request, operation, bucketName, returnFuture, bucketRegion);
+                    } else {
+                        returnFuture.completeExceptionally(throwable);
+                    }
+                } else {
+                    returnFuture.completeExceptionally(throwable);
+                }
+            } else {
+                CompletableFuture<ReturnT> newFuture = operation.apply(request);
+                CompletableFutureUtils.forwardResultTo(newFuture, returnFuture);
+                CompletableFutureUtils.forwardExceptionTo(returnFuture, newFuture);
+            }
+        });
     }
 
-    //TODO: add cross region logic
-    static final class BucketEndpointProvider implements S3EndpointProvider {
-        private final S3EndpointProvider delegate;
-        private final String bucket;
-
-        private BucketEndpointProvider(S3EndpointProvider delegate, String bucket) {
-            this.delegate = delegate;
-            this.bucket = bucket;
-        }
-
-        public static BucketEndpointProvider create(S3EndpointProvider delegate, String bucket) {
-            return new BucketEndpointProvider(delegate, bucket);
-        }
+    private <T extends S3Request, ReturnT> void sendRequestWithRightRegion(T request,
+                                                                           Function<T, CompletableFuture<ReturnT>> operation,
+                                                                           String bucketName,
+                                                                           CompletableFuture<ReturnT> returnFuture,
+                                                                           Optional<String> bucketRegionFromException) {
+        String region = bucketRegionFromException.get();
+        bucketToRegionCache.put(bucketName, Region.of(region));
+        doSendRequestWithRightRegion(request, operation, returnFuture, region);
```

Comment:
In this particular case, I think the code is more readable if the contents of `doSendRequestWithRightRegion` is inlined

Reply:
done